torch/testing/_asserts.py

```diff
@@ -245,118 +264,350 @@ def _check_values_close(
+    """Amends an exception message.
+
+    Args:
+        exc (E): Exception.
+        msg_fmtstr: Format string for the amended message.
+
+    Returns:
+        (E): New exception with amended error message.
+    """
+    return type(exc)(msg_fmtstr.format(str(exc)))
+
+
+_SEQUENCE_MSG_FMTSTR = "The failure occurred at index {} of the sequences."
+
+
+def _check_sequence(
+    a: Sequence[Tensor], b: Sequence[Tensor], check_tensors: Callable[[Tensor, Tensor], Optional[Exception]]
+) -> Optional[Exception]:
+    """Checks if the values of two sequences of tensors match.
+
+    Args:
+        a (Sequence[Tensor]): First sequence of tensors.
+        b (Sequence[Tensor]): Second sequence of tensors.
+        check_tensors (Callable[[Tensor, Tensor], Optional[Exception]]): Callable used to check if the items of
+            :attr:`a` and :attr:`b` match. In case they mismatch should return an :class:`Exception` with an expressive
+            error message.
+
+    Returns:
+        Optional[Exception]: :class:`AssertionError` if the sequences do not have the same length. Additionally, any
+            exception returned by :attr:`check_tensors`. In this case, the error message is amended to include the
+            first offending index.
+    """
+    len_a = len(a)
+    len_b = len(b)
+    if len_a != len_b:
+        return AssertionError(f"The length of the sequences mismatch: {len_a} != {len_b}")
+    for idx, (a_t, b_t) in enumerate(zip(a, b)):
+        exc = check_tensors(a_t, b_t)
+        if exc:
+            return _amend_error_message(exc, f"{{}}\n\n{_SEQUENCE_MSG_FMTSTR.format(idx)}")
+
+    return None
+
+
+_MAPPING_MSG_FMTSTR = "The failure occurred for key '{}' of the mappings."
+
+
+def _check_mapping(
+    a: Mapping[Any, Tensor], b: Mapping[Any, Tensor], check_tensors: Callable[[Tensor, Tensor], Optional[Exception]]
+) -> Optional[Exception]:
+    """Checks if the values of two mappings of tensors match.
+
+    Args:
+        a (Mapping[Any, Tensor]): First mapping of tensors.
+        b (Mapping[Any, Tensor]): Second mapping of tensors.
+        check_tensors (Callable[[Tensor, Tensor], Optional[Exception]]): Callable used to check if the values of
+            :attr:`a` and :attr:`b` match. In case they mismatch should return an :class:`Exception` with an expressive
+            error message.
+
+    Returns:
+        Optional[Exception]: :class:`AssertionError` if the sequences do not have the same set of keys. Additionally,
+            any exception returned by :attr:`check_tensors`. In this case, the error message is amended to include the
+            first offending key.
+    """
+    a_keys = set(a.keys())
+    b_keys = set(b.keys())
+    if a_keys != b_keys:
+        missing_keys = b_keys - a_keys
+        additional_keys = a_keys - b_keys
+        return AssertionError(
+            f"The keys of the mappings do not match:\n\n"
+            f"Missing keys in the first mapping: {sorted(missing_keys)}\n"
```

Comment:
This is cool and we shouldn't block this diff on this question, but this might be more natural with "expected" and "actual" vs. "first"

Reply:
Resolves automatically with https://github.com/pytorch/pytorch/pull/55145#discussion_r613042365